### PR TITLE
losslesscut-bin: init at 3.33.1

### PR DESCRIPTION
**Pull request by ShamrockLee**
Add binary package for
* Linux (AppImage)
* Mac (dmg, x86_64 version)
* Windows (zip)

###### Motivation for this change
If applied, [LosslessCut](https://github.com/mifi/lossless-cut) -- a convenient video editor -- will be available in nixpkgs for Linux, Mac and Windows users.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/video/lossless-cut-bin/linuxbin.nix [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# Binary currently ususable (but is build successfully).
```
Comment (Mindavi) [inline pkgs/applications/video/lossless-cut-bin/linuxbin.nix:1]: Is this statement still true?

**Review by Mindavi — COMMENTED** [review]
Haven't looked too closely to the package definition, but building and running on Linux NixOS works. Cutting out a simple video also works.

I wonder; is there a reason as to not build the electron app from source?

File: pkgs/applications/video/lossless-cut-bin/linuxbin.nix (hunk 1 above)
Comment (ShamrockLee) [reply pkgs/applications/video/lossless-cut-bin/linuxbin.nix:1]: I tried but it doesn't work.
I'll remove it.

**Comment by ShamrockLee** [comment]
Thanks for review and reply.
> I wonder; is there a reason as to not build the electron app from source?

It would be wonderful to build from source, and that's what I have tried at first (#91683).

Firstly, I personally have never succeeded in the source-build of any electron project.
They requires development dependencies to work, which means an electron node package will be downloaded and installed by the package manager. The problem is that it always want to download something at build phase as part of the "installation" of the electron package.

Secondly, the author of LosslessCut decided to bundle statically-build ffmpeg and ffprobe libreary instead of relying on the one provided by the package manager ar the environment, and the bundle will be downloaded during installation. We need to hack on the code to make it use the one downloaded by fetchurl or the one linked from the nixpkgs derivation.

It would be great to get some advice on how to build them.

**Comment by Mindavi** [comment]
I understand. I've never built an electron package before, so I can't really give any guidance on that.

IMO this can be merged (I don't have commit rights).

**Comment by ShamrockLee** [comment]
Some modifications:
  * Use `lib` instead of `stdenv.lib` to fit #108938.
  * Bug fix on the source URL. I had forgotten to change the owner name when copy-paste the expression from another work of mine.

**Review by veprbl — APPROVED** [review]
Runs on macOS

File: pkgs/applications/video/lossless-cut-bin/windows.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+  installPhase = ''
+    mkdir -p $out/bin
+    mkdir -p $out/libexec
+    cd ..
+    mv ${nameSourceBase} $out/libexec
+
+  '' + (getSymlinkCommand "${nameSourceBase}/${nameExecutable}" "$out/bin/${nameExecutable}");
+
+  meta.platforms = with lib.platforms; windows;
```
Comment (SuperSandro2000) [inline pkgs/applications/video/lossless-cut-bin/windows.nix:45]:
```suggestion
  meta.platforms = lib.platforms.windows;
```

File: pkgs/applications/video/lossless-cut-bin/windows.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+  unpackPhase = ''
+    unzip $src -d ${nameSourceBase}
+  '';
+
+  sourceRoot = nameSourceBase;
+
+  installPhase = ''
+    mkdir -p $out/bin
+    mkdir -p $out/libexec
```
Comment (SuperSandro2000) [inline pkgs/applications/video/lossless-cut-bin/windows.nix:39]:
```suggestion
    mkdir -p $out/bin $out/libexec
```

File: pkgs/applications/video/lossless-cut-bin/windows.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  nameSourceBase = "${nameCamel}-win";
+  nameSource = "${nameSourceBase}.zip";
+  nameExecutable = "${nameCamel}.exe";
+  owner = "mifi";
+  getSymlinkCommand = if (customSymlinkCommand != null) then customSymlinkCommand
+    else if useMklink then (targetPath: linkPath: "mklink ${targetPath} ${linkPath}")
+    else (targetPath: linkPath: "ln -s ${targetPath} ${linkPath}");
+in stdenvNoCC.mkDerivation {
+
+  inherit pname version;
```
Comment (SuperSandro2000) [inline pkgs/applications/video/lossless-cut-bin/windows.nix:21]:
```suggestion
in stdenvNoCC.mkDerivation {
  inherit pname version;
```

File: pkgs/applications/video/lossless-cut-bin/dmg.nix [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  nameApp = nameCamel + ".app";
+  owner = "mifi";
+  src = fetchurl {
+    url = "https://github.com/${owner}/${pname}/releases/download/v${version}/${nameSource}";
+    name = nameSource;
+    sha256 = "1z7h6mwb6dwdj9hy5kfpxc065nn81fbq860afrl1r8xq26rpz2wp";
+  };
+in stdenvNoCC.mkDerivation {
+
+  inherit pname version src;
```
Comment (SuperSandro2000) [inline pkgs/applications/video/lossless-cut-bin/dmg.nix:16]:
```suggestion
in stdenvNoCC.mkDerivation {
  inherit pname version src;
```

File: pkgs/applications/video/lossless-cut-bin/dmg.nix [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    undmg ${src}
+  '';
+  sourceRoot = nameApp;
+
+  installPhase = ''
+    mkdir -p $out/Applications/${nameApp}
+    cp -R . $out/Applications/${nameApp}
+  '';
+
+  meta.platforms = with lib.platforms; darwin;
```
Comment (SuperSandro2000) [inline pkgs/applications/video/lossless-cut-bin/dmg.nix:30]:
```suggestion
  meta.platforms = lib.platforms.darwin;
```

File: pkgs/applications/video/lossless-cut-bin/appimage.nix [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    mkdir $out/share/applications
+    cp ${extracted}/${nameExecutable}.desktop $out/share/applications
+    substituteInPlace $out/share/applications/${nameExecutable}.desktop \
+        --replace AppRun ${nameExecutable}
+  '';
+
+  meta.platforms = with lib.platforms;
+    (lib.lists.intersectLists x86_64 (
+      linux ++ freebsd ++ netbsd ++ openbsd
+    ));
```
Comment (SuperSandro2000) [inline pkgs/applications/video/lossless-cut-bin/appimage.nix:47]:
```suggestion
  meta.platforms = with lib.platforms; [ "x86_64-linux" ];
```

**Comment by ShamrockLee** [comment]
Rebase onto recent Nixpkgs master and upgrade to 3.33.1

**Comment by SuperSandro2000** [comment]
No idea if this is a bit overengineered.

**Comment by veprbl** [comment]
> No idea if this is a bit overengineered.

I think it could use some refactoring, but I would not say it's terribly complicated.

File: pkgs/applications/video/lossless-cut-bin/default.nix [hunk 8]
```diff
@@ -0,0 +1,31 @@
+{ pkgs, stdenvNoCC, lib }:
```
Comment (veprbl) [inline pkgs/applications/video/lossless-cut-bin/default.nix:1]:
```suggestion
{ stdenv, callPackage, lib }:
```

File: pkgs/applications/video/lossless-cut-bin/default.nix [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  passthru = {
+    appimage = callBin ./appimage.nix;
+    dmg = callBin ./dmg.nix;
+    windows = callBin ./windows.nix;
+  };
+  meta = with lib; {
+    description = "The swiss army knife of lossless video/audio editing";
+    homepage = "https://mifi.no/losslesscut/";
+    license = licenses.mit;
+    maintainers = with maintainers; [ ShamrockLee ];
```
Comment (veprbl) [inline pkgs/applications/video/lossless-cut-bin/default.nix:16]:
```suggestion
    description = "The swiss army knife of lossless video/audio editing";
    homepage = "https://mifi.no/losslesscut/";
    license = licenses.mit;
    maintainers = with maintainers; [ ShamrockLee ];
```
Why not put these into the override below, get rid of `callBin`

File: pkgs/applications/video/lossless-cut-bin/default.nix [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  };
+in (
+  if stdenvNoCC.isDarwin then passthru.dmg
```
Comment (veprbl) [inline pkgs/applications/video/lossless-cut-bin/default.nix:19]:
```suggestion
  if stdenvNoCC.isDarwin then callPackage ./dmg.nix { inherit version }
```
and so on

**Review by veprbl — COMMENTED** [review]
I would suggest to just merge this, but taking `pkgs` as attribute is not very acceptable.

**Comment by veprbl** [comment]
Also, why do you call the package "lossless-cut" if its "losslesscut"

**Comment by veprbl** [comment]
e.g. https://aur.archlinux.org/packages/losslesscut-bin/

**Comment by ShamrockLee** [comment]
> Also, why do you call the package "lossless-cut" if its "losslesscut"

It's just because the source repo is named "lossless-cut", and the manual suggests that the attribute be identical to the project name. (or do I misunderstand something?)

`losslesscut` sounds better, and is also the name of the command to get it start.

**Comment by ShamrockLee** [comment]
Apply the suggestions and rename

**Comment by ryneeverett** [comment]
Seems like calling the per-platform files in a let-expression makes them impossible to override (AFAIK).

**Comment by ShamrockLee** [comment]
> Seems like calling the per-platform files in a let-expression makes them impossible to override (AFAIK).

@ryneeverett Would it be better for the three of them be visible in all-packages.nix or should I put them in the `passthru`?

**Comment by ryneeverett** [comment]
@ShamrockLee I'm not the person to ask but I would think passthru would be acceptable.

**Pull request merged**